sw/device/silicon_creator/lib/manifest.h

```diff
@@ -0,0 +1,210 @@
+static_assert(offsetof(manifest_t, image_length) == 392, "");
+static_assert(offsetof(manifest_t, image_version) == 396, "");
+static_assert(offsetof(manifest_t, image_timestamp) == 400, "");
+static_assert(offsetof(manifest_t, exponent) == 408, "");
+static_assert(offsetof(manifest_t, reserved1) == 412, "");
+static_assert(offsetof(manifest_t, usage_constraints) == 416u, "");
+static_assert(offsetof(manifest_t, lockdown_info) == 448u, "");
+static_assert(offsetof(manifest_t, modulus) == 464u, "");
+static_assert(offsetof(manifest_t, extension0_offset) == 848u, "");
+static_assert(offsetof(manifest_t, extension0_checksum) == 852u, "");
+static_assert(offsetof(manifest_t, extension1_offset) == 856u, "");
+static_assert(offsetof(manifest_t, extension1_checksum) == 860u, "");
+static_assert(offsetof(manifest_t, extension2_offset) == 864u, "");
+static_assert(offsetof(manifest_t, extension2_checksum) == 868u, "");
+static_assert(offsetof(manifest_t, extension3_offset) == 872u, "");
+static_assert(offsetof(manifest_t, extension3_checksum) == 876u, "");
+static_assert(sizeof(manifest_t) == 880, "");
+
+/**
+ * Helper macro for defining the manifest of a boot stage stored in flash.
+ *
+ * This definition places the manifest struct in the `.manifest` section, which
+ * is the first section of boot stages stored in flash, e.g. see
+ * sw/device/silicon_creator/rom_exts/rom_ext_common.ld. The `used` attribute
+ * ensures that the compiler emits the definition even if it is not used.
+ *
+ * Boot stages should not need to access their own manifests but since the
+ * manifest is updated using the signer tool, the manifest struct is
+ * volatile-qualified in case this assumption changes in the future.
+ */
+#define DEFINE_FLASH_BOOT_STAGE_MANIFEST(name)                                \
+  static const volatile manifest_t name __attribute__((section(".manifest"))) \
+      __attribute__((used))
+
+/**
+ * Signed region of an image.
+ */
+typedef struct manifest_signed_region {
+  /**
+   * Start of the signed region of an image.
+   */
+  const void *start;
+  /**
+   * Length of the signed region of an image in bytes.
+   */
+  size_t length;
+} manifest_signed_region_t;
+
+/**
+ * Gets the signed region of an image.
+ *
+ * This function also performs a basic check to ensure that the length of the
+ * image is within a reasonable range.
+ *
+ * @param manifest A manifest
+ * @param[out] signed_region Signed region of an image.
+ * @return The result of the operation.
+ */
+inline rom_error_t manifest_signed_region_get(
+    const manifest_t *manifest, manifest_signed_region_t *signed_region) {
+  // FIXME: We should also have a lower bound.
+  enum { kMaxLength = 64 * 1024 };
+  if (manifest->image_length > kMaxLength) {
+    return kErrorManifestInternal;
+  }
+  *signed_region = (manifest_signed_region_t){
+      .start = (const void *)&manifest->image_length,
+      .length = manifest->image_length - offsetof(manifest_t, image_length),
+  };
```

Comment:
This is probably nothing, but I'd have expected this to look like:
```
signed_region->start = ...;
signed_region->length = ...;
```

I don't have a problem with the way this is written per-se, but lets just double check the generated assembly.  The literal interpretation of this code is "create an anonymous manifest_signed_region_t and assign (ie: memcpy) it to signed_region".

(Really, this is probably nothing, but I've encountered compilers doing stupid things in the past).

Reply:
Note that this is semantically distinct, because it zeroes the other fields. However, I really doubt Clang will do anything different here. Since signed_region isn't volatile Clang can assign things however it pleases...

Reply:
The generated assembly is identical in both cases, the only difference in `mask_rom_sim_verilator.dis` for these two cases are debug sections and source code:

```
--- mask_rom_sim_verilator.compound_literal.dis 2021-06-01 16:26:00.034591103 -0400
+++ mask_rom_sim_verilator.individual_members.dis       2021-06-01 16:26:40.306540016 -0400
@@ -17,21 +17,21 @@
                   ALLOC
   6 .riscv.attributes 00000026  00000000  00000000  00003008  2**0
                   CONTENTS, READONLY
-  7 .debug_info   0000753c  00000000  00000000  0000302e  2**0
+  7 .debug_info   00007540  00000000  00000000  0000302e  2**0
                   CONTENTS, READONLY, DEBUGGING
-  8 .debug_abbrev 0000176a  00000000  00000000  0000a56a  2**0
+  8 .debug_abbrev 0000175d  00000000  00000000  0000a56e  2**0
                   CONTENTS, READONLY, DEBUGGING
-  9 .debug_aranges 00000080  00000000  00000000  0000bcd4  2**0
+  9 .debug_aranges 00000080  00000000  00000000  0000bccb  2**0
                   CONTENTS, READONLY, DEBUGGING
- 10 .debug_line   00004172  00000000  00000000  0000bd54  2**0
+ 10 .debug_line   00004172  00000000  00000000  0000bd4b  2**0
                   CONTENTS, READONLY, DEBUGGING
- 11 .debug_loc    00003cbf  00000000  00000000  0000fec6  2**0
+ 11 .debug_loc    00003cdc  00000000  00000000  0000febd  2**0
                   CONTENTS, READONLY, DEBUGGING
- 12 .debug_ranges 000002c0  00000000  00000000  00013b85  2**0
+ 12 .debug_ranges 000002c0  00000000  00000000  00013b99  2**0
                   CONTENTS, READONLY, DEBUGGING
- 13 .debug_str    00003f22  00000000  00000000  00013e45  2**0
+ 13 .debug_str    00003f22  00000000  00000000  00013e59  2**0
                   CONTENTS, READONLY, DEBUGGING
- 14 .debug_frame  000005e0  00000000  00000000  00017d68  2**2
+ 14 .debug_frame  000005e0  00000000  00000000  00017d7c  2**2
                   CONTENTS, READONLY, DEBUGGING
 
 Disassembly of section .vectors:
@@ -430,9 +430,9 @@
       manifest->image_length > kManifestImageLengthMax) {
     return kErrorManifestInternal;
   }
-  *signed_region = (manifest_signed_region_t){
-      .start = &manifest->image_length,
-      .length = manifest->image_length - offsetof(manifest_t, image_length),
+  signed_region->start = &manifest->image_length;
+  signed_region->length =
+      manifest->image_length - offsetof(manifest_t, image_length);
     8274:      e7858593                addi    a1,a1,-392
 mask_rom_boot():
 .../sw/device/silicon_creator/mask_rom/mask_rom.c:101
@@ -483,7 +483,7 @@
           (romextimage_entry_point *)manifest_entry_point_address_get(manifest);
     82a0:      5532                    lw      a0,44(sp)
 manifest_entry_point_address_get():
-.../sw/device/silicon_creator/lib/manifest.h:176
+.../sw/device/silicon_creator/lib/manifest.h:175
  * @return The entry point.
  */
 inline uintptr_t manifest_entry_point_address_get(const manifest_t *manifest) {
```